packages/palette/src/elements/Shelf/ShelfScrollBar.tsx

```diff
@@ -0,0 +1,201 @@
+import { themeGet } from "@styled-system/theme-get"
+import React, { useEffect, useRef, useState } from "react"
+import styled from "styled-components"
+import { Box, BoxProps } from "../Box"
+import { Clickable } from "../Clickable"
+import { useClickScroll } from "./useClickScroll"
+import { useDragScroll } from "./useDragScroll"
+
+interface ShelfScrollBarProps extends BoxProps {
+  viewport?: HTMLDivElement | null
+}
+
+/**
+ * A synthetic scrollbar
+ */
+export const ShelfScrollBar: React.FC<ShelfScrollBarProps> = React.memo(
+  ({ viewport, ...rest }) => {
+    const [
+      { scrollLeft, scrollWidth, clientWidth },
+      setScrollState,
+    ] = useState<ScrollState>({
+      scrollLeft: viewport?.scrollLeft ?? 0,
+      scrollWidth: viewport?.scrollWidth ?? 1,
+      clientWidth: viewport?.clientWidth ?? 1,
+    })
+
+    const trackRef = useRef<HTMLDivElement | null>(null)
+    const thumbRef = useRef<HTMLButtonElement | null>(null)
+
+    const trackWidth = trackRef.current?.clientWidth ?? 1
+
+    const {
+      progress,
+      requiresScrolling,
+      thumbOffset,
+      thumbWidth,
+    } = buildScrollBar({ trackWidth, scrollLeft, scrollWidth, clientWidth })
+
+    // Update scrollState on scroll and resize
+    useEffect(() => {
+      if (!viewport) return
+
+      const updateScrollState = () => {
+        setScrollState({
+          scrollLeft: viewport?.scrollLeft ?? 0,
+          scrollWidth: viewport?.scrollWidth ?? 1,
+          clientWidth: viewport?.clientWidth ?? 1,
+        })
+      }
+
+      updateScrollState()
+
+      viewport.addEventListener("scroll", updateScrollState, { passive: true })
+      window.addEventListener("resize", updateScrollState)
+
+      return () => {
+        viewport.removeEventListener("scroll", updateScrollState)
+        window.removeEventListener("resize", updateScrollState)
+      }
+    }, [viewport])
+
+    useDragScroll({
```

Comment:
I moved these different scroll behaviors into hooks just for organizational purposes. This was getting pretty long/confusing.